### PR TITLE
Kubemark scripts retry failed gcloud commands

### PR DESCRIPTION
**Pull request by mwielgus**
Continuation of #19992 by @gmarek.

### CODE REVIEW

**Comment by mwielgus** [comment]
Added "". Comments will have to be added by @gmarek in some next PR.

File: test/kubemark/common.sh [hunk 1]
```diff
@@ -23,3 +24,24 @@ export PROJECT
...
+RETRIES=3
+
+# arguments:
+# $@: all stuff that goes after 'gcloud compute '
+function run-gcloud-compute {
+  local attempt=0
+  while true; do
+    if ! gcloud compute $@; then
+      if (( $attempt > $RETRIES )); then
+        echo -e "${color_red} Failed to $1 $2 $3.${color_norm}" >& 2
```
Comment (ihmccreery) [inline test/kubemark/common.sh]: I'd rather print out `$@`, since that's what we're actually running.

File: test/kubemark/common.sh [hunk 2]
```diff
@@ -23,3 +24,24 @@ export PROJECT
 MASTER_NAME="${INSTANCE_PREFIX}-kubemark-master"
 MASTER_TAG="kubemark-master"
 EVENT_STORE_NAME="${INSTANCE_PREFIX}-event-store"
+
+RETRIES=3
+
+# arguments:
+# $@: all stuff that goes after 'gcloud compute '
+function run-gcloud-compute {
```
Comment (ihmccreery) [inline test/kubemark/common.sh]: `run-gcloud-compute-with-retries`?

File: test/kubemark/common.sh [hunk 3]
```diff
@@ -23,3 +24,24 @@ export PROJECT
 MASTER_NAME="${INSTANCE_PREFIX}-kubemark-master"
 MASTER_TAG="kubemark-master"
 EVENT_STORE_NAME="${INSTANCE_PREFIX}-event-store"
+
+RETRIES=3
+
+# arguments:
```
Comment (ihmccreery) [inline test/kubemark/common.sh]: Please add a comment explaining what this does & why.

File: test/kubemark/start-kubemark.sh [hunk 4]
```diff
@@ -46,47 +46,47 @@ cd $CURR_DIR
...
-    --project "${PROJECT}" \
-    --network "${NETWORK}" \
-    --source-ranges "0.0.0.0/0" \
-    --target-tags "${MASTER_TAG}" \
-    --allow "tcp:443"
+run-gcloud-compute disks create "${MASTER_NAME}"-pd \
+  ${GCLOUD_COMMON_ARGS} \
+  --type ${MASTER_DISK_TYPE} \
+  --size ${MASTER_DISK_SIZE}
+
```
Comment (ihmccreery) [inline test/kubemark/start-kubemark.sh]: Generally avoid whitespace like this.

File: test/kubemark/start-kubemark.sh [hunk 5]
```diff
@@ -46,47 +46,47 @@ cd $CURR_DIR
...
+run-gcloud-compute instances create "${MASTER_NAME}" \
+  ${GCLOUD_COMMON_ARGS} \
+  --machine-type ${MASTER_SIZE} \
+  --image-project=${MASTER_IMAGE_PROJECT} \
+  --image ${MASTER_IMAGE} \
+  --tags ${MASTER_TAG} \
+  --network ${NETWORK} \
+  --scopes storage-ro,compute-rw,logging-write \
+  --disk name=${MASTER_NAME}-pd,device-name=master-pd,mode=rw,boot=no,auto-delete=no
+
```
Comment (ihmccreery) [inline test/kubemark/start-kubemark.sh]: Avoid whitespace.

File: test/kubemark/start-kubemark.sh [hunk 6]
```diff
@@ -46,47 +46,47 @@ cd $CURR_DIR
...
-    --scopes "storage-ro,compute-rw,logging-write" \
-    --disk "name=${MASTER_NAME}-pd,device-name=master-pd,mode=rw,boot=no,auto-delete=no"
-
-gcloud compute firewall-rules create "${INSTANCE_PREFIX}-kubemark-master-https" \
-    --project "${PROJECT}" \
-    --network "${NETWORK}" \
-    --source-ranges "0.0.0.0/0" \
-    --target-tags "${MASTER_TAG}" \
-    --allow "tcp:443"
+run-gcloud-compute disks create "${MASTER_NAME}"-pd \
```
Comment (ihmccreery) [inline test/kubemark/start-kubemark.sh]: `"${MASTER_NAME}-pd"`?

File: test/kubemark/start-kubemark.sh [hunk 7]
```diff
@@ -46,47 +46,47 @@ cd $CURR_DIR
...
+  ${GCLOUD_COMMON_ARGS} \
```
Comment (ihmccreery) [inline test/kubemark/start-kubemark.sh]: This should be in double quotes, no?  And below; why did all the quotes get removed?

**Comment by ihmccreery** [comment]
A lot of these changes are just removing double quotes, which goes against Bash style.  Why?

File: test/kubemark/common.sh (hunk 1 above)
Comment (mwielgus) [inline test/kubemark/common.sh]: Done

File: test/kubemark/common.sh (hunk 2 above)
Comment (mwielgus) [inline test/kubemark/common.sh]: Done

File: test/kubemark/common.sh (hunk 3 above)
Comment (mwielgus) [inline test/kubemark/common.sh]: Done

File: test/kubemark/start-kubemark.sh (hunk 7 above)
Comment (mwielgus) [inline test/kubemark/start-kubemark.sh]: Done

File: test/kubemark/start-kubemark.sh (hunk 4 above)
Comment (mwielgus) [inline test/kubemark/start-kubemark.sh]: Done

File: test/kubemark/start-kubemark.sh (hunk 6 above)
Comment (mwielgus) [inline test/kubemark/start-kubemark.sh]: Done

File: test/kubemark/start-kubemark.sh (hunk 7 above)
Comment (mwielgus) [inline test/kubemark/start-kubemark.sh]: No, why?

File: test/kubemark/start-kubemark.sh (hunk 5 above)
Comment (mwielgus) [inline test/kubemark/start-kubemark.sh]: Done

**Comment by davidopp** [comment]
@ihmccreery it looks like @mwielgus pushed a new commit -- could you take another look?

**Comment by davidopp** [comment]
ping

**Comment by ihmccreery** [comment]
> A lot of these changes are just removing double quotes, which goes against Bash style. Why?

This doesn't seem to have been addressed; the [Shell Style Guide](https://google.github.io/styleguide/shell.xml#Quoting) says:

> Always quote strings containing variables, command substitutions, spaces or shell meta characters, unless careful unquoted expansion is required.

**Comment by mwielgus** [comment]
@ihmccreery Added back the quote strings.

File: test/kubemark/start-kubemark.sh [hunk 8]
```diff
@@ -46,39 +46,39 @@ cd $CURR_DIR
...
+  --machine-type "${MASTER_SIZE}" \
+  --image-project="${MASTER_IMAGE_PROJECT}" \
+  --image "${MASTER_IMAGE}" \
+  --tags "${MASTER_TAG}" \
+  --network "${NETWORK}" \
+  --scopes "storage-ro,compute-rw,logging-write" \
+  --disk "name=${MASTER_NAME}-pd,device-name=master-pd,mode=rw,boot=no,auto-delete=no"
+
+run-gcloud-compute-with-retries firewall-rules create ${INSTANCE_PREFIX}-kubemark-master-https \
```
Comment (ihmccreery) [inline test/kubemark/start-kubemark.sh]: `"${INSTANCE_PREFIX}-kubemark-master-https"`

File: test/kubemark/start-kubemark.sh [hunk 9]
```diff
@@ -46,39 +46,39 @@ cd $CURR_DIR
...
+  --target-tags "${MASTER_TAG}" \
+  --allow "tcp:443"
 
 MASTER_IP=$(gcloud compute instances describe ${MASTER_NAME} \
   --zone="${ZONE}" --project="${PROJECT}" | grep natIP: | cut -f2 -d":" | sed "s/ //g")
 
 if [ "${SEPARATE_EVENT_MACHINE:-false}" == "true" ]; then
   EVENT_STORE_NAME="${INSTANCE_PREFIX}-event-store"
-  gcloud compute disks create "${EVENT_STORE_NAME}-pd" \
+    run-gcloud-compute-with-retries disks create ${EVENT_STORE_NAME}-pd \
```
Comment (ihmccreery) [inline test/kubemark/start-kubemark.sh]: `"${EVENT_STORE_NAME}-pd"`

File: test/kubemark/start-kubemark.sh [hunk 10]
```diff
@@ -46,39 +46,39 @@ cd $CURR_DIR
...
       ${GCLOUD_COMMON_ARGS} \
       --type "${MASTER_DISK_TYPE}" \
       --size "${MASTER_DISK_SIZE}"
 
-  gcloud compute instances create "${EVENT_STORE_NAME}" \
+    run-gcloud-compute-with-retries instances create ${EVENT_STORE_NAME} \
```
Comment (ihmccreery) [inline test/kubemark/start-kubemark.sh]: `"${EVENT_STORE_NAME}"`

File: test/kubemark/common.sh [hunk 11]
```diff
@@ -23,3 +24,26 @@ export PROJECT
 MASTER_NAME="${INSTANCE_PREFIX}-kubemark-master"
 MASTER_TAG="kubemark-master"
 EVENT_STORE_NAME="${INSTANCE_PREFIX}-event-store"
+
+RETRIES=3
+
+# Runs gcloudn compute command with the given parameters. Up to $RETRIES will be made
```
Comment (ihmccreery) [inline test/kubemark/common.sh]: s/gcloudn/gcloud

File: test/kubemark/common.sh [hunk 12]
```diff
@@ -23,3 +24,26 @@ export PROJECT
...
+# to execute the command.
+# arguments:
+# $@: all stuff that goes after 'gcloud compute '
+function run-gcloud-compute-with-retries {
+  local attempt=0
+  while true; do
```
Comment (ihmccreery) [inline test/kubemark/common.sh]: Consider using something like

```
for retry in $(seq 1 ${RETRIES}); do
```

See examples in cluster/gce/util.sh.  I think that should simplify the logic a bit.

**Comment by ihmccreery** [comment]
@mwielgus Made comments.

File: test/kubemark/start-kubemark.sh (hunk 8 above)
Comment (mwielgus) [inline test/kubemark/start-kubemark.sh]: done

File: test/kubemark/start-kubemark.sh (hunk 9 above)
Comment (mwielgus) [inline test/kubemark/start-kubemark.sh]: done

File: test/kubemark/start-kubemark.sh (hunk 10 above)
Comment (mwielgus) [inline test/kubemark/start-kubemark.sh]: done

File: test/kubemark/common.sh (hunk 12 above)
Comment (mwielgus) [inline test/kubemark/common.sh]: done

File: test/kubemark/common.sh (hunk 11 above)
Comment (mwielgus) [inline test/kubemark/common.sh]: done

**Comment by mwielgus** [comment]
Applied comments.

**Comment by ihmccreery** [comment]
LGTM, thanks @mwielgus!

**Comment by fabioy** [comment]
Manually merged.

**Comment by davidopp** [comment]
ref/ #18334

**Comment by davidopp** [comment]
This broke kubernetes-kubemark-gce e2e. I'll send a revert.

http://kubekins.dls.corp.google.com/view/Critical%20Builds/job/kubernetes-kubemark-gce/4969/console

**Comment by gmarek** [comment]
The problem was the `return` stmt. IIUC it returns the value of last thing that finished. In our case it was `! gcloud compute ...` thus it returned 'false'. This is exactly the reason why I wasn't a fan of writing the `for` loop there.

Reopened #19992 with the fix.

**Comment by gmarek** [comment]
Uhg. I wasn't able to reopen the PR. I created a new one: #20401

**Pull request merged**